### PR TITLE
nvidia-drivers: refactor & nvidia_x11 -> nvidia-drivers

### PR DESCRIPTION
**Pull request by codyopel**
- Support for version 355+
  - NVIDIA has added experimental support for EGL, bringing wayland support
    closer (does not work with wayland yet)
- Fix i686 support (for whatever thats worth)
  - Cuda UVM is only supported on x86_64 for 346+
- Rename nvidia_x11 -> nvidia-drivers
  - The NVIDIA driver package contains drivers for several things beyond X.Org
    so referring to it as an X11 driver is inaccurate.
  - Updated references to reflect these change
    - While updating these references it appears several packages are not
      linking against nvidia-drivers correcty, in that they use a specific
      version. Better seperation of kernelspace & userspace is required to
      solve conflicts with some packages.
- Use the long-lived driver as the default
  - Currently the short lived driver is used which may contain breaking changes.
- Add short-lived driver
  - Available as "nvidiaUnstable" in services.xserver.videoDrivers
- Remove support for 173.xx & 304.xx
  - These driver branches require versions of the kernel or xorg-server that are
    no longer supported in nixpkgs.
- Fix patchelf --shrink-rpath
- Install everything explicitly to catch build failures due to changes
- Cleanup shared object version symlinking, _.so_
- Cuda fixes
  - Add missing cuda executables and manpages
- GTK+3 support and used by default (GTK+3 is supported for 346+)
  - nvidia-settings now works correctly on high dpi displays
- Add additional seperation of kernelspace and userspace to fix packaging of
  packages building against the nvidia-drivers. This is currently imcomplete,
  but the ground work for doing so is complete.
- Add nvidia default application profiles
- Cleanup buildinputs

cc @edolstra @vcunat 

All branches have been tested, as well as 32-bit lib support on x86_64.

Versions 173.xx & 304.xx are removed because they are ABI incompatible with
xorg-server 1.17, which is the current version in nixpkgs.

### CODE REVIEW

File: pkgs/os-specific/linux/nvidia-drivers/default.nix [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+  disallowedReferences =
+    if buildKernelspace then
+      [ kernel.dev ]
+    else
+      [ ];
+
+  patches =
+    optionals (versionMajor == "340") [
+      #./nvidia-340.76-kernel-4.0.patch
```
Comment (vcunat) [inline pkgs/os-specific/linux/nvidia-drivers/default.nix]: Why have the patch and reference without actually using it?

File: pkgs/os-specific/linux/nvidia-drivers/default.nix [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        ]
+      )
+    )
+  );
+
+  passthru = {
+    inherit
+      version
+      versionMajor;
+    nvenc = (
```
Comment (vcunat) [inline pkgs/os-specific/linux/nvidia-drivers/default.nix]: Nitpick: I would personally remove parentheses around these two `if`s in `passthru`. They shouldn't be needed and even visually it's expressed by indentation.

File: pkgs/os-specific/linux/nvidia-drivers/default.nix [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+, libsOnly ? false
+
+, nvidiasettingsSupport ? true
+  , atk ? null
+  , gdk_pixbuf ? null
+  , glib ? null
+  , pango ? null
+  , gtk2Support ? false
+    , gtk2 ? null
+  , gtk3Support ? true
```
Comment (vcunat) [inline pkgs/os-specific/linux/nvidia-drivers/default.nix]: I remember Eelco didn't like the dependency on gtk3 by default (a few months ago). I think we should start preferring gtk3 at *some* point &ndash; even in general though I don't think we need a mass switch in cases where both versions are well supported.

**Comment by vcunat** [comment]
Niptick we can fix during cherry-picking: there should be an empty line after the first one in the commit message.

**Comment by codyopel** [comment]
@vcunat I updated it with 304.xx, I haven't fully tested it though so the kernel module will need patching with newer kernels.

**Comment by vcunat** [comment]
Why did you close this?

**Pull request closed without merging**